### PR TITLE
Rewrite requiring components and plugins

### PR DESCRIPTION
**Pull request by fobos**
Removed slot symlink to itself on node_modeules. Symlink are creating during `npm install` process in postinstall script.

Rewrite functions that load `components` and `plugins` into end-user application.

### CODE REVIEW

**Comment by fobos** [comment]
Strictly recommended to review by @ovvost, @heilage-nsk and @myshov

File: app/index.js [hunk 1]
```diff
@@ -742,3 +709,56 @@ if (DEBUG) {
...
+/**
+ * Requires some specific entity by its location and name.
+ * Seeks the entity in end-user application then seeks it in the framework.
+ *
+ * @param {string} location The entity location
+ * @param {string} name Module name to load
+ * @return {*}
+ */
+function lookupRequire(location, name) {
+    var modulePath = location + '/' + name;
```
Comment (fobos) [inline app/index.js]: Guys, please make attention to such concat. I think it is enough here right now. But looks not so solid.

Comment (heilage-nsk) [inline app/index.js]: require('path').join(location, name) or will fail under windows

File: app/index.js [hunk 2]
```diff
@@ -372,22 +349,12 @@ Application.prototype.notify = function(moduleId, message) {
...
-    // Сначала пытаемся загрузить из приложения, если там нет - из слота
-    try {
-        component = this.require('components/' + name + '/' + name);
-    } catch (e) {
-        component = require('slot/components/' + name);
-    }
-
-    return component;
-};
+Application.prototype.loadComponent = _.partial(lookupRequire, 'components');
```
Comment (heilage-nsk) [inline app/index.js]: Consider adding _.memoize here to cache loaded component.

File: app/index.js [hunk 3]
```diff
@@ -742,3 +709,56 @@ if (DEBUG) {
...
+    var modulePath = location + '/' + name;
+    var res;
+
+    try { // Require from end-user app
+        res = env.require(modulePath + '/' + name);
+    } catch (e) { // Require from framework
+        if (this.isClient) { // Require via browserify in end-user app
+            res = require('slot/' + modulePath);
+        } else {
+            res = require('../' + modulePath);
```
Comment (heilage-nsk) [inline app/index.js]: Looks like kludge.
However, I see no problem if it works.

**Comment by myshov** [comment]
postinstall.js is not deleted

**Comment by ovvost** [comment]
Delete `postinstall.js` and :+1:

File: app/index.js [hunk 4]
```diff
@@ -742,3 +709,56 @@ if (DEBUG) {
...
+ * @return {*}
+ */
+function lookupRequire(location, name) {
+    var modulePath = location + '/' + name;
+    var res;
+
+    try { // Require from end-user app
+        res = env.require(modulePath + '/' + name);
+    } catch (e) { // Require from framework
+        if (this.isClient) { // Require via browserify in end-user app
```
Comment (heilage-nsk) [inline app/index.js]: check if `this` works fine inside of _.partial/_.memoize

Comment (fobos) [inline app/index.js]: Wow! Look like my big mistake during refactoring. It won't work correct right now. Big thnx

File: app/index.js (hunk 3 above)
Comment (fobos) [inline app/index.js]: Unfortunately yes, but more pretty than symlink. I know correct way to solve it but later

File: app/index.js (hunk 1 above)
Comment (fobos) [inline app/index.js]: Unfortunately we have tons of code that doesn't support windows :( Won't fix right now.

Comment (ktoto) [inline app/index.js]: it's must be ok. I can do `require('slot/plugins/xx')` on Windows and all be required properly.

**Comment by fobos** [comment]
I'm testing the solution on staging of `2gis.ru` project before merging the PR

File: app/index.js [hunk 5]
```diff
@@ -742,3 +709,56 @@ if (DEBUG) {
...
+/**
+ * Setup plugins.
+ *
+ * @param {Array} plugins list of plugins
+ * @private
+ */
+Application.prototype._setupPlugins = function(plugins) {
+    _.each(plugins, function(name) {
+        var pluginModule = lookupRequire('plugins', name);
+        this[name] = pluginModule(this);
```
Comment (ktoto) [inline app/index.js]: try/catch? if plugin initialization will be broken.

**Comment by ilyapro** [comment]
:+1:

**Comment by ktoto** [comment]
:+1:

**Comment by myshov** [comment]
:+1:

**Comment by ilyapro** [comment]
!smoke

**Comment by fobos** [comment]
So all tests on `2gis.ru` have been passed.

**Pull request merged**